tests/NuGet.Services.AzureSearch.Tests/VersionListsFacts.cs

```diff
@@ -119,6 +122,312 @@ public void ReturnsCurrentSetOfVersions()
+                //     [ D-1.0.0, L-2.0.0 ], [ D-1.0.0, U-2.0.0 ], [ D-1.0.0, D-2.0.0 ]
+                //   ]
+                //
+                // The initialState sequence is this full enumeration without the Deleted version action. This is
+                // because a package deleted initially will simply not be present. The changesToApply sequence is this
+                // full enumeration with all version actions (as in the example).
+                //
+                // Finally, the cartesian produce of these two sequences is produced. Each pair is a test case.
+                var allActions = Enum.GetValues(typeof(VersionAction)).Cast<VersionAction>().ToArray();
+                var actionsExceptDelete = allActions.Where(x => x != VersionAction.Deleted).ToArray();
+
+                foreach (var initialState in GetAllVersionListChanges(fullVersions, actionsExceptDelete))
+                {
+                    foreach (var changeToApply in GetAllVersionListChanges(fullVersions, allActions))
+                    {
+                        yield return new TestCase(initialState.ToArray(), changeToApply.ToList());
+                    }
+                }
+            }
+
+            private static IEnumerable<IEnumerable<VersionListChange>> GetAllVersionListChanges(
+                IReadOnlyCollection<string> fullVersion,
+                IReadOnlyList<VersionAction> actions)
+            {
+                foreach (var versionSubsetSequence in SubsetsOf(fullVersion))
+                {
+                    var versionSubset = versionSubsetSequence.ToList();
+                    var combinations = CombinationsOfTwo(versionSubset, actions);
+                    foreach (var combination in combinations)
+                    {
+                        yield return combination.Select(x => ToVersionListChange(x.Item1, x.Item2));
+                    }
+                }
+            }
+
+            private static VersionListChange ToVersionListChange(string fullVersion, VersionAction action)
+            {
+                switch (action)
+                {
+                    case VersionAction.Listed:
+                        return VersionListChange.Upsert(
+                            fullVersion,
+                            new VersionPropertiesData(listed: true, semVer2: false));
+                    case VersionAction.Unlisted:
+                        return VersionListChange.Upsert(
+                            fullVersion,
+                            new VersionPropertiesData(listed: false, semVer2: false));
+                    case VersionAction.Deleted:
+                        return VersionListChange.Delete(fullVersion);
+                    default:
+                        throw new NotSupportedException($"The version action {action} is not supported.");
+                }
+            }
+
+            private static IEnumerable<IEnumerable<Tuple<T1, T2>>> CombinationsOfTwo<T1, T2>(
```

Comment:
>CombinationsOfTwo [](start = 67, length = 17)

Could you add mini examples on what `CombinationsOfTwo` and `CombinationsOfTwoByIndex` do?